receiver/bigipreceiver/client.go

```diff
@@ -0,0 +1,345 @@
+	var (
+		poolMembers         *models.PoolMembers
+		combinedPoolMembers *models.PoolMembers
+	)
+	errorReceived := false
+	collectedPoolMembers := false
+
+	// for each pool get pool member info and aggregate it into a single spot
+	for poolURL := range pools.Entries {
+		poolMemberPath := strings.TrimPrefix(poolURL, "https://localhost")
+		poolMemberPath = strings.TrimSuffix(poolMemberPath, "/stats") + poolMembersStatsPathSuffix
+
+		if err := c.get(ctx, poolMemberPath, &poolMembers); err != nil {
+			c.logger.Warn("Failed to retrieve all pool members", zap.Error(err))
+			errorReceived = true
+		} else {
+			combinedPoolMembers = combinePoolMembers(combinedPoolMembers, poolMembers)
+			collectedPoolMembers = true
+		}
+	}
+
+	if errorReceived && !collectedPoolMembers {
+		return nil, errCollectedNoPoolMembers
+	}
+
+	return combinedPoolMembers, nil
+}
+
+// GetNodes makes a call the statistics version of the nodes endpoint and returns the data.
+func (c *bigipClient) GetNodes(ctx context.Context) (*models.Nodes, error) {
+	var nodes *models.Nodes
+
+	if err := c.get(ctx, nodesStatsPath, &nodes); err != nil {
+		c.logger.Debug("Failed to retrieve nodes", zap.Error(err))
+		return nil, err
+	}
+
+	return nodes, nil
+}
+
+// post makes a POST request for the passed in path and stores result in the respObj
+func (c *bigipClient) post(ctx context.Context, path string, respObj interface{}) error {
+	// Construct endpoint and create request
+	url := c.hostEndpoint + path
+	postBody, _ := json.Marshal(map[string]string{
+		"username":          c.creds.username,
+		"password":          c.creds.password,
+		"loginProviderName": "tmos",
+	})
+	requestBody := bytes.NewBuffer(postBody)
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, requestBody)
+	if err != nil {
+		return fmt.Errorf("failed to create post request for path %s: %w", path, err)
+	}
+
+	// Make request
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to make http request: %w", err)
+	}
+
+	// Defer body close
+	defer func() {
+		if closeErr := resp.Body.Close(); closeErr != nil {
+			c.logger.Warn("failed to close response body", zap.Error(closeErr))
+		}
+	}()
+
+	// Check for OK status code
+	if resp.StatusCode != http.StatusOK {
+		c.logger.Debug("Big-IP API non-200", zap.Error(err), zap.Int("status_code", resp.StatusCode))
+
+		// Attempt to extract the error payload
+		payloadData, err := io.ReadAll(resp.Body)
+		if err != nil {
+			c.logger.Debug("failed to read payload error message", zap.Error(err))
+		} else {
+			c.logger.Debug("Big-IP API Error", zap.ByteString("api_error", payloadData))
+		}
+
+		return fmt.Errorf("non 200 code returned %d", resp.StatusCode)
+	}
+
+	// Decode the payload into the passed in response object
+	if err := json.NewDecoder(resp.Body).Decode(respObj); err != nil {
+		return fmt.Errorf("failed to decode response payload: %w", err)
+	}
```

Comment:
This code is almost identical to the get method below. Can we generalize this a bit or remove the get/post methods in favor of a more generalize request method?

Reply:
@cpheps good call out. i was planning on refactoring this at some point but then forgot about it

Reply:
@cpheps changed